projects/npm-tools/packages/npm-scripts/src/jsp/getTaglibInvocations.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * SPDX-FileCopyrightText: © 2019 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const Lexer = require('./Lexer');
+
+const DEFAULT_OPTIONS = {
+	ELEnabled: true,
+};
+
+function lex(source, options = {}) {
+	const {ELEnabled} = {
+		...DEFAULT_OPTIONS,
+		...options,
+	};
+
+	const lexer = new Lexer((api) => {
+		const {consume, fail, match, meta, peek, repeat, sequence, token} = api;
+
+		meta.set('ELEnabled', !!ELEnabled);
+		meta.commit();
+
+		const {
+			ATTRIBUTE,
+			CHAR,
+			CUSTOM_ACTION,
+			CUSTOM_ACTION_START,
+			SPACE,
+		} = require('./jspDSL')(api);
+
+		return () => {
+			if (peek(sequence(CUSTOM_ACTION_START, CUSTOM_ACTION))) {
+				consume(CUSTOM_ACTION_START);
+
+				return token('TAG', consume(CUSTOM_ACTION));
+			}
+			else if (peek(SPACE)) {
+				consume(SPACE);
+
+				return token('SPACE', '·');
+			}
+			else if (peek(ATTRIBUTE)) {
+				return token('ATTRIBUTE', consume());
+			}
+			else if (peek(match('>'))) {
+				consume(repeat(CHAR));
+
+				return token('REST', '·');
+			}
+			else {
+				fail('Failed to consume all input');
+			}
+		};
+	});
+
+	return [...lexer.lex(source)];
+}
+
+module.exports = function (jsp, ...tags) {
+	const invocations = [];
+
+	for (const tag of tags) {
+		let from = 0;
+
+		while ((from = jsp.indexOf(`<${tag}`, from)) != -1) {
+			let to = jsp.indexOf(`</${tag}>`, from + 1 + tag.length);
+
+			if (to === -1) {
+				to = jsp.indexOf('/>', from + 1 + tag.length);
+
+				if (to === -1) {
+					break;
+				}
+			}
+			else {
+				to += 3 + tag.length;
+			}
+
+			const jspFragment = jsp.substring(from, to);
```

Comment:
So here I'm obviously thinking about nested tags. Now given the tags that you're passing in from `TAGLIBS`, none of those can ever be nested recursively, can they? I don't think they can, but just wanted to check with you.

Reply:
No. They cannot be nested.

Also take into account that this linter won't be able to cover all cases, so it's just a "best effort" type of check. We are leaving out JS modules inside Java and dynamic JS modules, for example. So, I don't think it makes sense to put excessive effort in this feature as the benefit will be insignificant compared to the complexity of the task. I would say handling the happy path is enough...

Reply:
Also, I prefer to miss some valid cases rather than showing false positives :grimacing: . It's not this case, but just wanted to explain which where my assumptions when writing this feature...